File: src/util/pid.py
class PID:
    def __init__(self, k_p, k_d, k_i, target, init_value=0.0):
        self.Kp = k_p
        self.Kd = k_d
        self.Ki = k_i
        self.target = target
        self.lastValue = init_value
        self.integral_error = 0.0

    def get_control(self, value, dt):
        """
            Returns PID control.
            Derivative spike is mitigated by dError/dt = -dInput/dt
            see: http://brettbeauregard.com/blog/2011/04/improving-the-beginner%e2%80%99s-pid-derivative-kick/
        """
        error = self.target - value

        derivative = -(value - self.lastValue) / dt

        self.lastValue = value
        self.integral_error += error * dt
        return self.Kp * error + self.Kd * derivative + self.Ki * self.integral_error

    def set_target(self, target):
        self.target = target

File: esp32/balancing-robot/src/pid/PID.cpp
#include "PID.h"

PID::PID(float Kp, float Kd, float Ki, float target) {
    _Kp = Kp;
    _Kd = Kd;
    _Ki = Ki;
    _target = target;
}

float PID::getControl(float value, float dt_seconds) {
    float lastValue = _has_last_value ? _last_value : value;
    float error = _target - value;
    float de = -(value - lastValue) / dt_seconds;
    _integralError += _Ki * error * dt_seconds;
    _lastError = error;
    _last_value = value;
    _has_last_value = true;
    return (_Kp * error + _Kd * de + _integralError);
}

void PID::setSettings(float Kp, float Kd, float Ki) {
  _Kp = Kp;
  _Kd = Kd;
  _Ki = Ki;
}

void PID::setTarget(float target) {
    _target = target;
    _integralError = .0;
}